File: plugins/argo-data-analyst/skills/run-analysis/lib/python/argo_analysis/figures.py
#!/usr/bin/env python3
"""argo_analysis.figures — two charts, one look, print-ready.

    from argo_analysis import figures
    figures.bar_by_group(study, "education", "redcap_data_access_group",
                         "outputs/education_by_site.png")
    figures.hist(study, "age", "outputs/age.png")

THE LOOK, AND WHY IT IS FIXED
-----------------------------
Both charts come out the same way, because a figure in a paper is read at 8 cm
wide by someone who is not looking for it:

  * PNG at 300 dpi — the resolution journals ask for, big enough to crop
  * the Okabe-Ito palette, which stays distinguishable for readers with any of
    the common forms of colour blindness, and still separates in greyscale print
  * the title is the field's question, in the words the participant was asked
  * the subtitle is always `n = … ; missing = …`, because a chart whose
    denominator you cannot see is a chart you cannot check
  * a footnote naming the script and the date — the provenance of a figure is
    the script that made it, and a figure that has drifted from the file it came
    from is how the wrong version ends up in a submission
  * no chartjunk: no top or right frame, a faint horizontal grid, nothing else

Percentages here obey the same rule as everywhere in this library: they are out
of the records the field was actually asked of and that answered it.

If matplotlib is not installed, nothing crashes. Both functions say so in one
sentence, name the one command that fixes it, and return None — the table part
of an analysis is worth having even on a laptop that cannot draw.
"""

from __future__ import annotations

import datetime
import importlib
import sys
from pathlib import Path

try:
    from . import core, table1 as _table1
except ImportError:                   # run as a loose file from its own folder
    import core                       # type: ignore
    import table1 as _table1          # type: ignore


#: Okabe-Ito: eight colours chosen to stay apart for colour-blind readers.
#: Ordered so the first two — blue and orange — are the pair used most often.
PALETTE = ["#0072B2", "#E69F00", "#009E73", "#CC79A7",
           "#56B4E9", "#D55E00", "#F0E442", "#000000"]

DPI = 300
FIGSIZE = (7.5, 4.5)
GRID_COLOUR = "#D9D9D9"
INK = "#222222"
QUIET_INK = "#666666"

MATPLOTLIB_MISSING = (
    "This chart needs matplotlib, a free add-on for Python that draws graphs.\n"
    "It is not installed on this computer yet, so the chart was skipped — the rest\n"
    "of the analysis is unaffected. To draw it, open a terminal window, type the\n"
    "line below, press Enter, wait for it to finish, then run this again:\n\n"
    "    python3 -m pip install matplotlib\n"
)


def _pyplot():
    """matplotlib's drawing surface, or None with one plain paragraph of explanation.

    Imported fresh each time rather than cached at the top of the file: this
    module has to be importable, and its docstring readable, on a machine where
    matplotlib was never installed.
    """
    try:
        matplotlib = importlib.import_module("matplotlib")
        matplotlib.use("Agg")             # no screen needed; write straight to a file
        return importlib.import_module("matplotlib.pyplot")
    except ImportError:
        print(MATPLOTLIB_MISSING, file=sys.stderr)
        return None


def _dress(fig, ax, title, subtitle):
    """The house look, applied in one place so both charts cannot drift apart."""
    ax.set_title("")
    fig.suptitle(title, x=0.01, y=0.98, ha="left", fontsize=13,
                 fontweight="bold", color=INK)
    fig.text(0.01, 0.90, subtitle, ha="left", fontsize=9.5, color=QUIET_INK)
    fig.text(0.01, 0.015, f"{_generating_script()} — {datetime.date.today().isoformat()}",
             ha="left", fontsize=7, color=QUIET_INK)
    for side in ("top", "right"):
        ax.spines[side].set_visible(False)
    for side in ("left", "bottom"):
        ax.spines[side].set_color(GRID_COLOUR)
    ax.yaxis.grid(True, color=GRID_COLOUR, linewidth=0.8)
    ax.xaxis.grid(False)
    ax.set_axisbelow(True)
    ax.tick_params(colors=QUIET_INK, labelsize=9.5, length=0)


def _generating_script() -> str:
    argv0 = sys.argv[0] if sys.argv else ""
    if not argv0 or argv0 in ("-c", "-"):
        return "an interactive session"
    return Path(argv0).name


def _save(fig, path, plt):
    path = Path(path)
    path.parent.mkdir(parents=True, exist_ok=True)
    fig.savefig(path, dpi=DPI, bbox_inches="tight", facecolor="white")
    plt.close(fig)
    print(f"Chart written to {path}")
    return path


def _counts(study, field, mask):
    """{level label: count} for one column of records, in codebook order, and the
    number of records that were asked the field and answered it."""
    mapping = core.labels(study, field)
    applies = core.applicable(study, field)
    usable = core.usable(study, field)
    answered = mask & applies & usable
    out = {}
    if core.field_type(study, field) == "checkbox":
        for code, label in mapping.items():
            column = f"{field}___{code}"
            if column in study.data.columns:
                ticked = study.data[column].astype(str).str.strip() == "1"
                out[label] = int((mask & applies & ticked).sum())
            else:
                out[label] = 0
    else:
        values = study.data[field].astype(str).str.strip()
        for code, label in mapping.items():
            out[label] = int((answered & (values == code)).sum())
    return out, int(answered.sum()), int((mask & applies).sum()) - int(answered.sum())


def bar_by_group(study, field, group_by, path):
    """A grouped bar chart: one cluster per level of `field`, one bar per group.

    Bars are percentages within each group — the honest comparison when the
    groups are different sizes — out of the records that group was asked the
    field and answered it. The counts are printed above the bars so nothing is
    hidden behind a percentage.

    Returns the path written, or None if matplotlib is not installed.
    """
    plt = _pyplot()
    if plt is None:
        return None
    if group_by not in study.data.columns:
        raise ValueError(f"'{group_by}' is not a column in this export, so the chart "
                         "cannot be grouped by it.")
    mapping = core.labels(study, field)
    if not mapping:
        raise ValueError(f"'{field}' has no choice list in the codebook, so it has no bars "
                         "to draw. Use hist() for a number.")

    masks = _table1.group_masks(study, group_by)
    groups = [g for g in masks if g != "overall"]
    levels = list(mapping.values())

    per_group = {}
    for group in groups:
        counts, answered, _ = _counts(study, field, masks[group])
        per_group[group] = (counts, answered)
    _, answered_total, missing_total = _counts(study, field, masks["overall"])

    fig, ax = plt.subplots(figsize=FIGSIZE)
    width = 0.8 / max(len(groups), 1)
    for i, group in enumerate(groups):
        counts, answered = per_group[group]
        heights = [(100.0 * counts[l] / answered if answered else 0.0) for l in levels]
        positions = [x + i * width - 0.4 + width / 2 for x in range(len(levels))]
        bars = ax.bar(positions, heights, width=width * 0.92,
                      color=PALETTE[i % len(PALETTE)],
                      label=f"{group} (n = {answered})")
        for bar, level in zip(bars, levels):
            ax.annotate(str(counts[level]),
                        (bar.get_x() + bar.get_width() / 2, bar.get_height()),
                        ha="center", va="bottom", fontsize=7.5, color=QUIET_INK)

    ax.set_xticks(range(len(levels)))
    ax.set_xticklabels(levels, fontsize=9.5)
    ax.set_ylabel("% of those asked and answered", fontsize=9.5, color=QUIET_INK)
    if len(groups) > 1:
        ax.legend(frameon=False, fontsize=9, loc="upper right")
    _dress(fig, ax, core.field_label(study, field),
           f"n = {answered_total} ; missing = {missing_total}")
    fig.subplots_adjust(top=0.80)
    return _save(fig, path, plt)


def hist(study, field, path, bins: int = 20):
    """A histogram of one number, over everyone the field was asked of.

    Returns the path written, or None if matplotlib is not installed.
    """
    plt = _pyplot()
    if plt is None:
        return None
    applies = core.applicable(study, field)
    usable = core.usable(study, field)
    keep = applies & usable
    values = []
    for raw in study.data.loc[keep, field].astype(str).str.strip().tolist():
        try:
            values.append(float(raw))
        except (TypeError, ValueError):
            pass
    missing = int(applies.sum()) - len(values)
    if not values:
        raise ValueError(f"'{field}' has no numbers in it to chart — every value is blank, "
                         "a missing-data code, or not a number.")

    fig, ax = plt.subplots(figsize=FIGSIZE)
    ax.hist(values, bins=bins, color=PALETTE[0], edgecolor="white", linewidth=0.6)
    ax.set_xlabel(core.field_label(study, field), fontsize=9.5, color=QUIET_INK)
    ax.set_ylabel("Number of records", fontsize=9.5, color=QUIET_INK)
    _dress(fig, ax, core.field_label(study, field),
           f"n = {len(values)} ; missing = {missing}")
    fig.subplots_adjust(top=0.80)
    return _save(fig, path, plt)


def _main() -> int:
    print(__doc__.strip())
    print("\nThis file is part of the ARGO analysis library. It is not run on its own —")
    print("an analysis script imports it. See the run-analysis skill for how to start one.")
    return 0


if __name__ == "__main__":
    sys.exit(_main())

File: plugins/argo-data-analyst/skills/run-analysis/lib/R/argo_analysis/figures.R
# figures.R -- two figures, one look.
#
#     source("lib/R/argo_analysis/core.R")
#     source("lib/R/argo_analysis/figures.R")
#     bar_by_group(study, "sex", "redcap_data_access_group", "out/sex_by_site.png")
#     hist(study, "age", "out/age.png")
#
# THE HOUSE LOOK, so every ARGO figure is recognisable:
#   * PNG at 300 dpi -- sharp when a journal or a report prints it
#   * text sized to be readable at print size, not just on screen
#   * the Okabe-Ito palette, which stays distinguishable to colourblind readers
#     and in greyscale
#   * title  = the field's label, as the codebook words it
#   * subtitle = "n = ... ; missing = ..." -- a figure that hides its denominator
#     is a figure that will be misread
#
# Base graphics on purpose: no ggplot2, no install step. The script that made the
# figure is its provenance -- there is no metadata to lose.
#
# hist() shadows base R's hist() once this file is sourced. That is the contract
# name; the base version is still there as graphics::hist().

if (!exists("applicable", mode = "function")) {
  stop(paste0("Load the core of the library first:\n",
              '    source("lib/R/argo_analysis/core.R")'), call. = FALSE)
}

# Okabe-Ito: eight colours that stay apart for every common form of colour
# blindness. Blue first because a single-series chart should not be red.
ARGO_PALETTE <- c("#0072B2", "#E69F00", "#009E73", "#CC79A7",
                  "#56B4E9", "#D55E00", "#F0E442", "#666666")

ARGO_FIGURE_DPI <- 300

#' Open a 300-dpi PNG device, whichever backend this R has.
argo_open_png <- function(path, width = 8, height = 5) {
  dir <- dirname(path)
  if (nzchar(dir) && !dir.exists(dir)) dir.create(dir, recursive = TRUE, showWarnings = FALSE)
  type <- if (isTRUE(capabilities("cairo"))) "cairo"
          else if (isTRUE(capabilities("aqua"))) "quartz"
          else NULL
  ok <- tryCatch({
    if (is.null(type)) {
      grDevices::png(path, width = width, height = height, units = "in",
                     res = ARGO_FIGURE_DPI)
    } else {
      grDevices::png(path, width = width, height = height, units = "in",
                     res = ARGO_FIGURE_DPI, type = type)
    }
    TRUE
  }, error = function(e) FALSE)
  if (!ok) {
    stop(paste0("This copy of R cannot save PNG pictures -- it was built without the\n",
                "graphics support that needs. The tables will still work. If you need the\n",
                "figures, install R from https://cran.r-project.org (the standard build\n",
                "includes it) and run this again."), call. = FALSE)
  }
  graphics::par(mar = c(5.5, 5, 4.5, 1.5), family = "", cex.axis = 1.0,
                cex.lab = 1.1, las = 1)
  invisible(TRUE)
}

#' Title + "n = ... ; missing = ..." subtitle, drawn the same way on every figure.
argo_figure_titles <- function(title, n, missing) {
  graphics::title(main = title, cex.main = 1.35, font.main = 2, line = 2.6)
  graphics::mtext(sprintf("n = %d ; missing = %d", as.integer(n), as.integer(missing)),
                  side = 3, line = 1.0, cex = 1.0, col = "#444444")
}

#' Counts of a categorical field, side by side across the groups.
#'
#' Bars are the PERCENT within each group, because groups are rarely the same
#' size and raw counts invite the wrong comparison. The counts are in the table;
#' this picture is for the shape.
#'
#' Levels are in codebook order. Missing values and MDC codes are not a bar --
#' they are in the subtitle, where they belong.
bar_by_group <- function(study, field, group_by, path, width = 8, height = 5) {
  if (!(field %in% names(study$data))) {
    stop(sprintf("There is no column called '%s' in this export.", field), call. = FALSE)
  }
  if (!(group_by %in% names(study$data))) {
    stop(sprintf("There is no column called '%s' to group by.", group_by), call. = FALSE)
  }
  map <- labels(study, field)
  if (!length(map)) {
    stop(sprintf(paste0("'%s' has no choices in the codebook, so there are no bars to ",
                        "draw.\nFor a number, use hist(study, \"%s\", ...) instead."),
                 field, field), call. = FALSE)
  }
  codes <- names(map)[!(names(map) %in% ARGO_MDC)]

  applies <- applicable(study, field)
  values  <- argo_usable(study$data[[field]])
  gvals   <- trimws(as.character(study$data[[group_by]]))
  gvals[is.na(gvals)] <- ""

  levels_ <- argo_group_levels_for_figures(study, group_by, gvals)
  gmap    <- argo_choice_map(study, group_by)
  gnames  <- vapply(levels_, function(lv)
    if (lv %in% names(gmap) && nzchar(gmap[[lv]])) unname(gmap[[lv]]) else lv, character(1))

  pct <- matrix(0, nrow = length(codes), ncol = length(levels_),
                dimnames = list(unname(map[codes]), unname(gnames)))
  for (j in seq_along(levels_)) {
    keep <- applies & gvals == levels_[j]
    v <- values[keep]
    v <- v[!is.na(v)]
    if (length(v)) {
      for (i in seq_along(codes)) pct[i, j] <- 100 * sum(v == codes[i]) / length(v)
    }
  }

  n_applicable <- sum(applies)
  n_missing    <- sum(applies & is.na(values))

  argo_open_png(path, width, height)
  on.exit(grDevices::dev.off(), add = TRUE)
  colours <- rep(ARGO_PALETTE, length.out = nrow(pct))
  # Fit the axis to the data, rounded up to a tidy multiple of 10, never past
  # 100 and never below 20 -- a bar chart of 2% against a 0-100 axis is unreadable,
  # and one against a 0-3 axis is a lie.
  top <- min(100, max(20, ceiling(max(pct) * 1.2 / 10) * 10))
  bp <- graphics::barplot(pct, beside = TRUE, col = colours, border = NA,
                          ylim = c(0, top), ylab = "Percent of participants (%)",
                          xlab = argo_field_label(study, group_by),
                          cex.names = 1.0, axes = TRUE)
  graphics::abline(h = graphics::axTicks(2), col = "#EEEEEE", lwd = 1)
  graphics::barplot(pct, beside = TRUE, col = colours, border = NA, add = TRUE, axes = FALSE)
  graphics::legend("topright", legend = rownames(pct), fill = colours, border = NA,
                   bty = "n", cex = 0.95, inset = c(0, -0.02), xpd = TRUE)
  argo_figure_titles(argo_field_label(study, field), n_applicable - n_missing, n_missing)
  invisible(path)
}

# Group levels for a figure: codebook order where there is a codebook, else the
# order the data offers, sorted. Kept beside the figures so figures.R can be
# sourced without table1.R.
argo_group_levels_for_figures <- function(study, group_by, gvals) {
  observed <- unique(gvals[nzchar(gvals)])
  codebook <- names(argo_choice_map(study, group_by))
  if (length(codebook)) c(codebook[codebook %in% observed], sort(setdiff(observed, codebook)))
  else sort(observed)
}

#' The distribution of a numeric field.
#'
#' Missing values and MDC codes are excluded from the bars and reported in the
#' subtitle. The median is drawn as a dashed line, because the first question
#' anyone asks of a histogram is "where is the middle".
hist <- function(study, field, path, bins = 20, width = 8, height = 5) {
  if (!(field %in% names(study$data))) {
    stop(sprintf("There is no column called '%s' in this export.", field), call. = FALSE)
  }
  applies <- applicable(study, field)
  values  <- suppressWarnings(as.numeric(argo_usable(study$data[[field]])))
  v <- values[applies]
  n_missing <- sum(is.na(v))
  v <- v[!is.na(v)]
  if (!length(v)) {
    stop(sprintf(paste0("'%s' has no usable numbers in this export, so there is nothing ",
                        "to plot.\nIf it is a question with choices, use bar_by_group() ",
                        "instead."), field), call. = FALSE)
  }

  breaks <- if (is.null(bins)) "Sturges" else as.integer(bins)
  # Measure first, then draw: the extra headroom keeps the median legend off the
  # tallest bar.
  shape <- graphics::hist(v, breaks = breaks, plot = FALSE)
  argo_open_png(path, width, height)
  on.exit(grDevices::dev.off(), add = TRUE)
  graphics::hist(v, breaks = breaks, col = ARGO_PALETTE[1], border = "white",
                 main = "", xlab = argo_field_label(study, field),
                 ylab = "Number of participants",
                 ylim = c(0, max(shape$counts) * 1.18))
  graphics::abline(v = stats::median(v), col = ARGO_PALETTE[6], lwd = 2, lty = 2)
  graphics::legend("topright", legend = sprintf("median = %s", argo_format(stats::median(v), FALSE)),
                   col = ARGO_PALETTE[6], lwd = 2, lty = 2, bty = "n", cex = 0.95)
  argo_figure_titles(argo_field_label(study, field), length(v), n_missing)
  invisible(path)
}
